### PR TITLE
Create Series constructor for Array and Buffers

### PR DESCRIPTION
**Pull request by love-lena**
Related: #260

### CODE REVIEW

**Comment by love-lena** [comment]
I have added the Series constructors for Int32Arrays and Int32Buffers, and it works. Now I want to generalize it, and I am not sure of the best way to do that.

My list of things it should work for are:
- Int8Array
- Int16Array
- Int32Array
- Int64Array
- more

File: modules/cudf/src/series.ts [hunk 1]
```diff
@@ -202,6 +206,10 @@ export class AbstractSeries<T extends DataType = any> {
    * ```
    */
   static new<T extends DataType>(input: AbstractSeries<T>|Column<T>|SeriesProps<T>): Series<T>;
+
+  static new<T extends DataType>(input: Int32Array): Series<T>;
```
Comment (trxcllnt) [inline modules/cudf/src/series.ts:210]:
```suggestion
  static new(input: Int32Array): Series<Int32>;
```

File: modules/cudf/src/series.ts [hunk 2]
```diff
@@ -202,6 +206,10 @@ export class AbstractSeries<T extends DataType = any> {
...
+
+  static new<T extends DataType>(input: Int32Buffer): Series<T>;
```
Comment (trxcllnt) [inline modules/cudf/src/series.ts:212]:
```suggestion
  static new(input: Int32Buffer): Series<Int32>;
```

File: modules/cudf/src/series.ts [hunk 3]
```diff
@@ -1287,6 +1298,15 @@ function asColumn<T extends DataType>(
...
              {type: inferType(value), values: value as any, highWaterMark: Infinity})) as any;
   }
+  if (value instanceof Int32Array) {
+    return fromArrow(arrow.Vector.from(
+             {type: new Int32, values: value as any, highWaterMark: Infinity})) as any;
+  }
+  if (value instanceof Int32Buffer) {
+    const value_arr = new value.TypedArray(value);
+    return fromArrow(arrow.Vector.from(
+             {type: new Int32, values: value_arr as any, highWaterMark: Infinity})) as any;
```
Comment (trxcllnt) [inline modules/cudf/src/series.ts:1308]:
```suggestion
    return new Column({ type: new Int32, data: value, length: value.length });
```

File: modules/cudf/src/series.ts [hunk 4]
```diff
@@ -1287,6 +1298,15 @@ function asColumn<T extends DataType>(
     return fromArrow(arrow.Vector.from(
              {type: inferType(value), values: value as any, highWaterMark: Infinity})) as any;
   }
+  if (value instanceof Int32Array) {
+    return fromArrow(arrow.Vector.from(
+             {type: new Int32, values: value as any, highWaterMark: Infinity})) as any;
```
Comment (trxcllnt) [inline modules/cudf/src/series.ts:1303]:
```suggestion
    return new Column({ type: new Int32, data: value, length: value.length });
```

**Review by trxcllnt — CHANGES_REQUESTED** [review]
Left a few suggestions. Can simplify by returning a concrete Series since the inputs are host/device arrays of known byte-widths.

**Comment by bryevdv** [comment]
@trxcllnt would it be possible to make a generic function that extracts the input buffer/array type and maps it to the dtype, in order to collapse the number of explicit signatures and codepaths when more buffer/array types are added? I'm thinking similar to `ArrowToCUDFType` but perhaps thats not applicable in the same way?

**Comment by trxcllnt** [comment]
@bryevdv it is definitely possible, but from my experience trying similar things in Arrow, likely more trouble than it's worth

**Comment by love-lena** [comment]
I will work on a version thats just very explicit with the data then. For the 8-32 bit int types it wont be too much, if there are more Array/Buffer types we want to add overloads for it will get a little messier

File: modules/cudf/src/series.ts [hunk 5]
```diff
@@ -237,7 +268,9 @@ export class AbstractSeries<T extends DataType = any> {
    * const a = Series.new([1n, 2n, 3n, undefined, 4n]);
    * ```
    */
-  static new(input: (bigint|null|undefined)[]): Series<Int64>;
+  static new(input: (bigint|null|undefined)[]|BigInt64Array|Int64Buffer): Series<Int64>;
+  // static new(input: Float64Array|Float64Buffer): Series<Float64>;
```
Comment (love-lena) [inline modules/cudf/src/series.ts:272]: delete

File: modules/cudf/src/series.ts [hunk 6]
```diff
@@ -1207,7 +1242,7 @@ export {
   StructSeries,
 };
 
-function inferType(value: any[]): DataType {
+function inferType(value: any[]|TypedArray): DataType {
```
Comment (love-lena) [inline modules/cudf/src/series.ts:1245]: nope

File: modules/cudf/src/series.ts [hunk 7]
```diff
@@ -1269,39 +1304,87 @@ function inferType(value: any[]): DataType {
...
+//     case BigInt64Array: return new Int64;
+//     case Uint8Array: return new Uint8;
+//     case Uint16Array: return new Uint16;
+//     case Uint32Array: return new Uint32;
+//     case BigUint64Array: return new Uint64;
+//     case Float32Array: return new Float32;
+//     case Float64Array: return new Float64;
+//     default: break;
+//   }
+// }
```
Comment (love-lena) [inline modules/cudf/src/series.ts:1321]: why me

File: modules/cudf/src/series.ts [hunk 8]
```diff
@@ -1269,39 +1304,87 @@ function inferType(value: any[]): DataType {
...
+    return new Column({type: new Uint16, data: value, length: value.length});
+  } else if (value instanceof Uint32Array || value instanceof Uint32Buffer) {
+    return new Column({type: new Uint32, data: value, length: value.length});
+  } else if (value instanceof BigUint64Array || value instanceof Uint64Buffer) {
+    return new Column({type: new Uint64, data: value, length: value.length});
+  } else if (value instanceof Float32Array || value instanceof Float32Buffer) {
+    return new Column({type: new Float32, data: value, length: value.length});
+  } else if (value instanceof Float64Array || value instanceof Float64Buffer) {
+    return new Column({type: new Float64, data: value, length: value.length});
+  }
```
Comment (love-lena) [inline modules/cudf/src/series.ts:1377]: _|___|___|___|___|___|___|___|___|___|___|___|
___|___|___|___|___|___|___|___|___|___|___|_
_|___|___|___|___|___|___|___|___|___|___|___|
___|___|___|___|___|___|___|___|___|___|___|_
_|___|___|___|___|___|___|___|___|___|___|___|
___|___|___|___|___|___|___|___|___|___|___|_
_|___|___|___|___|___|___|___|___|___|___|___

Comment (love-lena) [inline modules/cudf/src/series.ts:1377]:
```
_|___|___|___|___|___|___|___|___|___|___|___|
___|___|___|___|___|___|___|___|___|___|___|_
_|___|___|___|___|___|___|___|___|___|___|___|
___|___|___|___|___|___|___|___|___|___|___|_
_|___|___|___|___|___|___|___|___|___|___|___|
___|___|___|___|___|___|___|___|___|___|___|_
_|___|___|___|___|___|___|___|___|___|___|___
```

Comment (trxcllnt) [reply modules/cudf/src/series.ts:1377]: Honestly this is fine with me. If you prefer, you can make it a switch-case (like below). Neither is particularly pleasing, but like I mentioned earlier today, probably not worth agonizing over.

```suggestion
  switch (value.constructor) {
    case Int8Buffer:
    case Int8Array: return new Column({type: new Int8, data: value});
    case Int16Buffer:
    case Int16Array: return new Column({type: new Int16, data: value});
    case Int32Buffer:
    case Int32Array: return new Column({type: new Int32, data: value});
    case Int64Buffer:
    case BigInt64Array: return new Column({type: new Int64, data: value});
    case Uint8Buffer:
    case Uint8Array: return new Column({type: new Uint8, data: value});
    case Uint16Buffer:
    case Uint16Array: return new Column({type: new Uint16, data: value});
    case Uint32Buffer:
    case Uint32Array: return new Column({type: new Uint32, data: value});
    case Uint64Buffer:
    case BigUint64Array: return new Column({type: new Uint64, data: value});
    case Float32Buffer:
    case Float32Array: return new Column({type: new Float32, data: value});
    case Float64Buffer:
    case Float64Array: return new Column({type: new Float64, data: value});
    default: break;
  }
```

**Review by trxcllnt — APPROVED** [review]

**Pull request merged**